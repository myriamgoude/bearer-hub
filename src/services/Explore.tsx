interface IItem {
  hubID: string
  title: string
}

interface ITemplate {
  hubID: string
  title: string
  provider: {
    hubID: string
    title: string
  }
}

// Returns a path to a Template page, given its Hub ID (a unique
// and required integer column) and title, and Hub ID and title
// of its provider
//
// E.g. "/integrations/5/slack-notification"
export function templatePath(template: ITemplate) {
  return `/integrations/${template.hubID}/${slug(template.title)}`
}

// Returns a path to a Provider page, given its Hub ID (a unique
// and required integer column) and title
//
// E.g. "/integrations/provider/2/slack/"
export function providerPath(provider: IItem) {
  return `/integrations/provider/${provider.hubID}/${slug(provider.title)}`
}

// Returns a path to a Category page, given its Hub ID (a unique
// and required integer column) and title
//
// E.g. "/integrations/category/3/communication/"
export function categoryPath(category: IItem) {
  return `/integrations/category/${category.hubID}/${slug(category.title)}`
}

// Generates a slug given an item with a title
//
// E.g. given item title "My Title", the generated slug will be
// "my-title"
export function slug(title: string) {
  return `${title.toLowerCase().replace(/\s/g, '-')}`
}

// Adds a given number of minutes to the current time
//
// Returns a String
export function timer(time: number) {
  const now = new Date()
  const minutesAdded = new Date(now.getTime() + time * 60 * 1000)
  const realTime = `${minutesAdded.getHours()}:${minutesAdded
    .getMinutes()
    .toString()
    .padStart(2, '0')}`
  return realTime
}
