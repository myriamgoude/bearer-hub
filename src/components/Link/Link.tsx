import React from 'react'
import { Link as GatsbyLink } from 'gatsby'
import { isBrowser } from '../../services/Browser'

interface ILinkProps {
  to: string
  trackLink?: boolean
  targetBlank?: boolean
  [key: string]: any
}

class Link extends React.Component<ILinkProps> {
  constructor(props: ILinkProps) {
    super(props)
  }

  // Typically the action (event name) is the internal link (e.g. /explore)
  // but this may be overridden and there are exceptions e.g. for some external links
  private getTrackingAction = () => {
    if (this.props.trackingAction) {
      return this.props.trackingAction
    }
    switch (this.props.to) {
      case 'https://docs.bearer.sh': {
        return 'general-documentation'
      }
      case '/integrations': {
        return 'explore-page'
      }
      case '/product': {
        return 'product-page'
      }
      default: {
        return this.props.to.replace(/\//g, '')
      }
    }
  }

  public handleOnClick = (event: any) => {
    // onClick action to track analytics
    if (this.props.trackLink && isBrowser) {
      const currentWindow = window as any
      currentWindow.analytics.track(this.getTrackingAction(), this.props.trackingOptions || {})
    }

    // other onClick actions (if present)
    if (this.props.onClick) {
      this.props.onClick(event)
    }
  }

  render() {
    const internal = /^\/(?!\/)/.test(this.props.to)
    const { to, trackLink, trackingAction, trackingOptions, onClick, targetBlank, ...remainingProps } = this.props

    if (internal) {
      return (
        <GatsbyLink
          to={this.props.to}
          {...(this.props.targetBlank ? { target: '_blank', rel: 'noopener' } : null)}
          {...(this.props.trackLink || this.props.onClick ? { onClick: this.handleOnClick } : null)}
          {...remainingProps}
        >
          {this.props.children}
        </GatsbyLink>
      )
    }
    return (
      <a
        href={this.props.to}
        {...(this.props.targetBlank ? { target: '_blank', rel: 'noopener' } : null)}
        {...(this.props.trackLink || this.props.onClick ? { onClick: this.handleOnClick } : null)}
        {...remainingProps}
      >
        {this.props.children}
      </a>
    )
  }
}

export default Link
