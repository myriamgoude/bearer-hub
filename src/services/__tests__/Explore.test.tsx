import React from 'react'
import MockDate from 'mockdate'
import * as ShallowRenderer from 'react-test-renderer/shallow'

import { templatePath, categoryPath, slug, timer } from '../Explore'
import Link from '../../components/Link/Link'

const myItem = { hubID: '123', title: 'My Item Title' }

describe('templatePath', () => {
  const myIntegration = { hubID: '456', title: 'My Integration Title', provider: myItem }
  const myPath = templatePath(myIntegration)

  it('returns the path to an integration presentation page', () => {
    expect(myPath).toMatch('/integrations/456/my-integration-title')
  })

  it('is considered an internal link', () => {
    // Note: Link will appear as mockConstructor in snapshot
    // since its mocked in /__mocks__/gatsby due to its dependacy on gatsby internals
    const renderer = ShallowRenderer.createRenderer()

    renderer.render(<Link to={myPath}>{myItem.title}</Link>)
    expect(renderer.getRenderOutput()).toMatchSnapshot()
  })
})

describe('categoryPath', () => {
  const myPath = categoryPath(myItem)

  it('returns the path to a category integrations page', () => {
    expect(myPath).toMatch('/integrations/category/123/my-item-title')
  })

  it('is considered an internal link', () => {
    // Note: Link will appear as mockConstructor in snapshot
    // since its mocked in /__mocks__/gatsby due to its dependacy on gatsby internals
    const renderer = ShallowRenderer.createRenderer()

    renderer.render(<Link to={myPath}>{myItem.title}</Link>)
    expect(renderer.getRenderOutput()).toMatchSnapshot()
  })
})

describe('slug', () => {
  it('returns a generated slug', () => {
    const mySlug = slug('My Dummy Title')
    expect(mySlug).toMatch('my-dummy-title')
  })
})

describe('timer', () => {
  afterEach(() => {
    MockDate.reset()
  })

  it('returns the expected time in 24:00 format', () => {
    MockDate.set(new Date('2019-01-01T12:01:11'))
    const myTime = timer(2)
    // 12:01 + 2 minutes = 12:03
    expect(myTime).toMatch('12:03')
  })

  it('returns the expected time across hours', () => {
    MockDate.set(new Date('2019-01-01T08:59:11'))
    const myTime = timer(2)
    // 8:59 + 2 minutes = 9:01
    expect(myTime).toMatch('9:01')
  })
})
