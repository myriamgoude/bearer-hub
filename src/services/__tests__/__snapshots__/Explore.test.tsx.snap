// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`categoryPath is considered an internal link 1`] = `
<mockConstructor
  to="/integrations/category/123/my-item-title"
>
  My Item Title
</mockConstructor>
`;

exports[`providerPath is considered an internal link 1`] = `
<mockConstructor
  to="/integrations/provider/123/my-item-title"
>
  My Item Title
</mockConstructor>
`;

exports[`templatePath is considered an internal link 1`] = `
<mockConstructor
  to="/integrations/456/my-integration-title"
>
  My Item Title
</mockConstructor>
`;
